In [116]:
#!/usr/bin/python


import pandas as pd
import sys
import pickle
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".
features_list = ['poi', 'bonus', 'deferral_payments','deferred_income', 'director_fees',
                 'exercised_stock_options','expenses', 'from_messages', 
                 'from_poi_to_this_person', 'from_this_person_to_poi',
                 'loan_advances', 'long_term_incentive', 'other',
                 'restricted_stock', 'restricted_stock_deferred',
                 'salary', 'shared_receipt_with_poi', 'to_messages',
                 'total_payments', 'total_stock_value'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)






In [117]:
data_dict.pop('TOTAL',0)

{'bonus': 97343619,
 'deferral_payments': 32083396,
 'deferred_income': -27992891,
 'director_fees': 1398517,
 'email_address': 'NaN',
 'exercised_stock_options': 311764000,
 'expenses': 5235198,
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 83925000,
 'long_term_incentive': 48521928,
 'other': 42667589,
 'poi': False,
 'restricted_stock': 130322299,
 'restricted_stock_deferred': -7576788,
 'salary': 26704229,
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 309886585,
 'total_stock_value': 434509511}

In [118]:
data_dict.pop('LOCKHART EUGENE E',0)

{'bonus': 'NaN',
 'deferral_payments': 'NaN',
 'deferred_income': 'NaN',
 'director_fees': 'NaN',
 'email_address': 'NaN',
 'exercised_stock_options': 'NaN',
 'expenses': 'NaN',
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 'NaN',
 'long_term_incentive': 'NaN',
 'other': 'NaN',
 'poi': False,
 'restricted_stock': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'salary': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 'NaN',
 'total_stock_value': 'NaN'}

In [119]:
data_dict.pop('THE TRAVEL AGENCY IN THE PARK',0)

{'bonus': 'NaN',
 'deferral_payments': 'NaN',
 'deferred_income': 'NaN',
 'director_fees': 'NaN',
 'email_address': 'NaN',
 'exercised_stock_options': 'NaN',
 'expenses': 'NaN',
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 'NaN',
 'long_term_incentive': 'NaN',
 'other': 362096,
 'poi': False,
 'restricted_stock': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'salary': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 362096,
 'total_stock_value': 'NaN'}

In [120]:
my_dataset = pd.DataFrame(data_dict)

There were 146 data points in this data set and now there are 143 data points.

In [121]:
my_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21 entries, bonus to total_stock_value
Columns: 143 entries, ALLEN PHILLIP K to YEAP SOON
dtypes: object(143)
memory usage: 23.6+ KB


In [122]:
my_dataset_explorer = my_dataset.transpose()

In [123]:
my_dataset_explorer.info()

<class 'pandas.core.frame.DataFrame'>
Index: 143 entries, ALLEN PHILLIP K to YEAP SOON
Data columns (total 21 columns):
bonus                        143 non-null object
deferral_payments            143 non-null object
deferred_income              143 non-null object
director_fees                143 non-null object
email_address                143 non-null object
exercised_stock_options      143 non-null object
expenses                     143 non-null object
from_messages                143 non-null object
from_poi_to_this_person      143 non-null object
from_this_person_to_poi      143 non-null object
loan_advances                143 non-null object
long_term_incentive          143 non-null object
other                        143 non-null object
poi                          143 non-null object
restricted_stock             143 non-null object
restricted_stock_deferred    143 non-null object
salary                       143 non-null object
shared_receipt_with_poi      143 non-null objec

There are 21 features including POI.

In [124]:
my_dataset_explorer.groupby('poi')[['poi']].count()

poi
poi       
False  125
True    18

There are only 18 POI's out of the 143 records

In [125]:
#remove text feature
email_df = my_dataset_explorer['email_address'].astype(str)
my_dataset_explorer = my_dataset_explorer.drop('email_address', axis = 1)

In [126]:
# Seperate the POI data to use for classifiers
poi_df = my_dataset_explorer['poi'].astype(int)
poi_df = pd.DataFrame(poi_df)
poi_df.head()


poi
ALLEN PHILLIP K       0
BADUM JAMES P         0
BANNANTINE JAMES M    0
BAXTER JOHN C         0
BAY FRANKLIN R        0

In [127]:
my_dataset_explorer = my_dataset_explorer.drop('poi', axis = 1)

In [128]:
my_dataset_explorer = my_dataset_explorer.replace('NaN',0.000000001)

In [129]:
my_dataset_explorer.apply(pd.to_numeric)

bonus  deferral_payments  deferred_income  \
ALLEN PHILLIP K              4.175000e+06       2.869717e+06    -3.081055e+06   
BADUM JAMES P                1.000000e-09       1.789800e+05     1.000000e-09   
BANNANTINE JAMES M           1.000000e-09       1.000000e-09    -5.104000e+03   
BAXTER JOHN C                1.200000e+06       1.295738e+06    -1.386055e+06   
BAY FRANKLIN R               4.000000e+05       2.604550e+05    -2.016410e+05   
BAZELIDES PHILIP J           1.000000e-09       6.846940e+05     1.000000e-09   
BECK SALLY W                 7.000000e+05       1.000000e-09     1.000000e-09   
BELDEN TIMOTHY N             5.249999e+06       2.144013e+06    -2.334434e+06   
BELFER ROBERT                1.000000e-09      -1.025000e+05     1.000000e-09   
BERBERIAN DAVID              1.000000e-09       1.000000e-09     1.000000e-09   
BERGSIEKER RICHARD P         2.500000e+05       1.000000e-09    -4.858130e+05   
BHATNAGAR SANJAY             1.000000e-09       1.000000e-09     1.000000e-09   
BIBI PHILIPPE A              1.000000e+06       1.000000e-09     1.000000e-09   
BLACHMAN JEREMY M            8.500000e+05       1.000000e-09     1.000000e-09   
BLAKE JR. NORMAN P           1.000000e-09       1.000000e-09    -1.137840e+05   
BOWEN JR RAYMOND M           1.350000e+06       1.000000e-09    -8.330000e+02   
BROWN MICHAEL                1.000000e-09       1.000000e-09     1.000000e-09   
BUCHANAN HAROLD G            5.000000e+05       1.000000e-09     1.000000e-09   
BUTTS ROBERT H               7.500000e+05       1.000000e-09    -7.500000e+04   
BUY RICHARD B                9.000000e+05       6.495840e+05    -6.948620e+05   
CALGER CHRISTOPHER F         1.250000e+06       1.000000e-09    -2.625000e+05   
CARTER REBECCA C             3.000000e+05       1.000000e-09    -1.597920e+05   
CAUSEY RICHARD A             1.000000e+06       1.000000e-09    -2.350000e+05   
CHAN RONNIE                  1.000000e-09       1.000000e-09    -9.878400e+04   
CHRISTODOULOU DIOMEDES       1.000000e-09       1.000000e-09     1.000000e-09   
CLINE KENNETH W              1.000000e-09       1.000000e-09     1.000000e-09   
COLWELL WESLEY               1.200000e+06       2.761000e+04    -1.440620e+05   
CORDES WILLIAM R             1.000000e-09       1.000000e-09     1.000000e-09   
COX DAVID                    8.000000e+05       1.000000e-09    -4.125000e+04   
CUMBERLAND MICHAEL S         3.250000e+05       1.000000e-09     1.000000e-09   
...                                   ...                ...              ...   
SAVAGE FRANK                 1.000000e-09       1.000000e-09    -1.212840e+05   
SCRIMSHAW MATTHEW            1.000000e-09       1.000000e-09     1.000000e-09   
SHANKMAN JEFFREY A           2.000000e+06       1.000000e-09     1.000000e-09   
SHAPIRO RICHARD S            6.500000e+05       1.000000e-09     1.000000e-09   
SHARP VICTORIA T             6.000000e+05       1.874690e+05     1.000000e-09   
SHELBY REX                   2.000000e+05       1.000000e-09    -4.167000e+03   
SHERRICK JEFFREY B           1.000000e-09       1.000000e-09     1.000000e-09   
SHERRIFF JOHN R              1.500000e+06       1.000000e-09     1.000000e-09   
SKILLING JEFFREY K           5.600000e+06       1.000000e-09     1.000000e-09   
STABLER FRANK                5.000000e+05       1.000000e-09     1.000000e-09   
SULLIVAN-SHAKLOVITZ COLLEEN  1.000000e+05       1.819930e+05     1.000000e-09   
SUNDE MARTIN                 7.000000e+05       1.000000e-09     1.000000e-09   
TAYLOR MITCHELL S            6.000000e+05       2.274490e+05     1.000000e-09   
THORN TERENCE H              1.000000e-09       1.658600e+04     1.000000e-09   
TILNEY ELIZABETH A           3.000000e+05       1.000000e-09    -5.750000e+05   
UMANOFF ADAM S               7.887500e+05       1.000000e-09     1.000000e-09   
URQUHART JOHN A              1.000000e-09       1.000000e-09    -3.666600e+04   
WAKEHAM JOHN                 1.000000e-09       1.000000e-09     1.

In [130]:
my_dataset_explorer.corrwith(poi_df['poi'])

bonus                        0.358486
deferral_payments           -0.039880
deferred_income             -0.274150
director_fees               -0.121886
exercised_stock_options      0.386853
expenses                     0.203545
from_messages               -0.034671
from_poi_to_this_person      0.189352
from_this_person_to_poi      0.128908
loan_advances                0.220183
long_term_incentive          0.256405
other                        0.169829
restricted_stock             0.247652
restricted_stock_deferred   -0.021548
salary                       0.338851
shared_receipt_with_poi      0.239625
to_messages                  0.107431
total_payments               0.242020
total_stock_value            0.382623
dtype: float64

The following have very little correlation with POI, so I will remove these features in addition to email_address.
restricted_stock_deferred   -0.021548
from_messages               -0.034671
deferral_payments           -0.039880

I now will have 17 features to explore.

I will run PCA on the six features that have the highest correlation:
exercised_stock_options      0.386853
total_stock_value            0.382623
bonus                        0.358486
salary                       0.338851
long_term_incentive          0.256405
deferred_income             -0.274150

In [131]:
features_list = ['poi', 'exercised_stock_options', 'total_stock_value', 'bonus', 
                 'salary', 'long_term_incentive', 'deferred_income'] 


In [132]:
my_dataset = my_dataset.replace('NaN',0.000000001)

In [133]:
my_dataset.loc['cash_payments'] = (my_dataset.loc['bonus']  
                                    + my_dataset.loc['salary']
                                    + my_dataset.loc['exercised_stock_options']
                                    + my_dataset.loc['director_fees'])

In [134]:
features_list = ['poi', 'exercised_stock_options', 'total_stock_value', 'bonus', 
                 'salary', 'long_term_incentive', 'deferred_income', 'cash_payments'] 

In [135]:
data = featureFormat(my_dataset, features_list, sort_keys=True)

In [136]:
labels, features = targetFeatureSplit(data)

In [137]:
from sklearn.cross_validation import train_test_split

features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)

In [138]:
from sklearn.preprocessing import MinMaxScaler
import math
import numpy as np

scaler = MinMaxScaler()
scaled_features_train = scaler.fit_transform(features_train)
scaled_features_test = scaler.fit_transform(features_test)

In [139]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, f1_score, accuracy_score, precision_score, confusion_matrix, recall_score

In [140]:
from sklearn.decomposition import PCA

In [141]:
pca = PCA(n_components = 3)
clf = GaussianNB()

In [142]:
from sklearn.model_selection import GridSearchCV, StratifiedShuffleSplit
sss = StratifiedShuffleSplit(100, test_size=0.5, random_state=42)


In [143]:
pipeline = Pipeline([("pca", pca), ("classifier", clf)])

In [144]:
param_grid = dict(pca__n_components=range(1,8))
                 

In [145]:
grid_search = GridSearchCV(pipeline, param_grid=param_grid, scoring='f1', cv = sss)

In [146]:
grid_fit = grid_search.fit(scaled_features_train, labels_train)

In [147]:
from sklearn.metrics import fbeta_score, make_scorer
from sklearn.model_selection import GridSearchCV

first_clf = grid_fit.best_estimator_

In [148]:
print first_clf

Pipeline(steps=[('pca', PCA(copy=True, iterated_power='auto', n_components=2, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('classifier', GaussianNB(priors=None))])


In [149]:
from sklearn.metrics import precision_recall_fscore_support
pred_train_gb = first_clf.predict(scaled_features_train)
gb_train_acc = accuracy_score(pred_train_gb, labels_train)
gb_train_all = precision_recall_fscore_support(pred_train_gb, labels_train)
pred_test_gb = first_clf.predict(scaled_features_test)
gb_test_acc = accuracy_score(pred_test_gb, labels_test)

In [150]:
from sklearn.ensemble import RandomForestClassifier

In [151]:
clf2 = RandomForestClassifier()

In [152]:
pipeline2 = Pipeline([("pca", pca), ("classifier", clf2)])

In [153]:
param_grid2 = dict(pca__n_components=range(1,8),
                 classifier__min_samples_split = (3, 50))

In [154]:
grid_search2 = GridSearchCV(pipeline2, param_grid=param_grid2, scoring='f1', cv = sss)


In [155]:
grid_fit2 = grid_search2.fit(scaled_features_train, labels_train)

In [156]:
second_clf = grid_fit2.best_estimator_

In [157]:
print second_clf

Pipeline(steps=[('pca', PCA(copy=True, iterated_power='auto', n_components=1, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('classifier', RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
     ...imators=10, n_jobs=1, oob_score=False, random_state=None,
            verbose=0, warm_start=False))])


In [158]:
pred_train_dt = second_clf.predict(scaled_features_train)
dt_train_acc = accuracy_score(pred_train_dt, labels_train)
dt_train_all = precision_recall_fscore_support(pred_train_dt, labels_train)
pred_test_dt = second_clf.predict(scaled_features_test)
dt_test_acc = accuracy_score(pred_test_dt, labels_test)

In [159]:
print ("gb_train_acc: "), gb_train_acc
print ("gb_test_acc: "), gb_test_acc
print ("gb_train_precision_recall_fscore_support: "), gb_train_all
print ("dt_train_acc: "), dt_train_acc
print ("dt_test_acc: "), dt_test_acc
print ("dt_train_precision_recall_fscore_support: "), dt_train_all

gb_train_acc:  0.88
gb_test_acc:  0.860465116279
gb_train_precision_recall_fscore_support:  (array([ 0.95402299,  0.38461538]), array([ 0.91208791,  0.55555556]), array([ 0.93258427,  0.45454545]), array([91,  9], dtype=int64))
dt_train_acc:  0.96
dt_test_acc:  0.767441860465
dt_train_precision_recall_fscore_support:  (array([ 0.97701149,  0.84615385]), array([ 0.97701149,  0.84615385]), array([ 0.97701149,  0.84615385]), array([87, 13], dtype=int64))


In [160]:
dump_classifier_and_data(first_clf, my_dataset, features_list)

In [161]:
from tester import test_classifier
test_classifier(first_clf, my_dataset, features_list)

Pipeline(steps=[('pca', PCA(copy=True, iterated_power='auto', n_components=2, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('classifier', GaussianNB(priors=None))])
	Accuracy: 0.86560	Precision: 0.49211	Recall: 0.24950	F1: 0.33112	F2: 0.27679
	Total predictions: 15000	True positives:  499	False positives:  515	False negatives: 1501	True negatives: 12485



I wonder if removing the feature that I engineered (cash_payments) would improve the scores.  Let's try it.

In [162]:
features_list = ['poi', 'exercised_stock_options', 'total_stock_value', 'bonus', 
                 'salary', 'long_term_incentive', 'deferred_income'] 

In [163]:
data = featureFormat(my_dataset, features_list, sort_keys=True)

In [164]:
labels, features = targetFeatureSplit(data)

In [165]:
from sklearn.cross_validation import train_test_split

features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)


In [166]:
scaler = MinMaxScaler()
scaled_features_train = scaler.fit_transform(features_train)
scaled_features_test = scaler.fit_transform(features_test)

In [167]:
pca3 = PCA(n_components = 3)
clf3 = GaussianNB()

In [168]:
pipeline3 = Pipeline([("pca", pca3), ("classifier", clf3)])

In [169]:
param_grid3 = dict(pca__n_components=range(1,7))

In [170]:
grid_search3 = GridSearchCV(pipeline3, param_grid=param_grid3, scoring='f1', cv = sss)

In [171]:
grid_fit3 = grid_search3.fit(scaled_features_train, labels_train)

In [172]:
third_clf = grid_fit.best_estimator_
print third_clf

Pipeline(steps=[('pca', PCA(copy=True, iterated_power='auto', n_components=2, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('classifier', GaussianNB(priors=None))])


In [173]:
dump_classifier_and_data(third_clf, my_dataset, features_list)

In [174]:
from tester import test_classifier
test_classifier(third_clf, my_dataset, features_list)

Pipeline(steps=[('pca', PCA(copy=True, iterated_power='auto', n_components=2, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)), ('classifier', GaussianNB(priors=None))])
	Accuracy: 0.87813	Precision: 0.57504	Recall: 0.32950	F1: 0.41894	F2: 0.36027
	Total predictions: 15000	True positives:  659	False positives:  487	False negatives: 1341	True negatives: 12513



Just as I suspected, removing my engineered feature improved the scores!